Track collapsable client events, that are handled optimistically (#444)

Alienjob opened the pull request:
Resolves (#444)




## Synopsis

Локально инициированные события обрабатываются оптимистично. Запросы на их обработку выполняются не последовательно. Может происходить ситуация, когда с бека приходит неактуальная мутация.




## Solution

При оптимистичной обработке события добавлять добавлять такое событие в специальный пул. Этот пул будет следить за тем, чтобы конкурирущие события были доставлены на бек последовательно. 

Так-же при получении с бека события следует проверять - было ли оно обработано оптимистично (хранится ли оно в пуле) и игнорировать такие события. Возможно от эту часть правильнее поручить беку - не присылать мутации по событиям, которые инициировала эта сессия.




## Checklist

- Created PR:
    - [x] In [draft mode][l:1]
    - [x] Name contains issue reference
    - [x] Has type and `k::` labels applied
- Before [review][l:4]:
    - [x] Documentation is updated (if required)
    - [x] Tests are updated (if required)
    - [x] Changes conform [code style][l:2]
    - [x] [CHANGELOG entry][l:3] is added (if required)
    - [x] FCM (final commit message) is posted or updated
    - [x] [Draft mode][l:1] is removed
- [ ] [Review][l:4] is completed and changes are approved
    - [ ] FCM (final commit message) is approved
- Before merge:
    - [ ] Milestone is set
    - [ ] PR's name and description are correct and up-to-date
    - [ ] All temporary labels are removed




[l:1]: https://help.github.com/en/articles/about-pull-requests#draft-pull-requests
[l:2]: /CONTRIBUTING.md#code-style
[l:3]: /CHANGELOG.md
[l:4]: https://help.github.com/en/articles/reviewing-changes-in-pull-requests

Alienjob committed:
init EventPool

Alienjob committed:
Track toggle user mute event

Alienjob committed:
Implement [EventPool] for [MyUserEvent]

Alienjob committed:
Implement lock on handler calls

Alienjob committed:
Update mute/unmute EventPool calls

Alienjob committed:
Implement neutralize and queue modes of events

Alienjob committed:
Implement event keys and ChatEvent extension

Alienjob committed:
Debug neutralized events

Alienjob committed:
Implement dependency injection for event pool

- rename EventsPool to OptimisticEventsPoolService and move to domain/service/optimistic_event_pool.dart
- add event pool to prorerties and constructors of ChatRepository, MyUserRepository, HiveRxChat
- update dependency injection in routes and tests

Alienjob commented:
@SleepySquash 
Добавлен сервис OptimisticEventsPoolService
он предоставляет единый алгоритм обработки оптимистичных событий
Когда пользователь добавляет чат в избранные,
1) событие обрабатывается и отражается в интерфейсе
2) в пул добавляется обработчик события, обработчики выполняются последовательно в рамках ключа события.

Ключ события в данном случае - тип события и ид пользователя. Если запросы по этому ключу выполнять параллельно. то они выполнятся в произвольном порядке и статус избранности на сервере может отличаться от статуса на клиенте. Выполнять последовательно вообще все запросы к серверу ухудшит производительность.

Сервер присылает событие. Возможно мы добавили чат в избранное в другом сеансе, возможно в этом. Событие игнорируется только если мы его отразили в интерфейсе и оно успело отправить запрос на сервер. Тогда оно попадает в список ожидающих игнорирования.

Если пользователь меняет статус 10 раз, пока первое событие еще не обработано сервером, нет смысла слать 10 событий на сервер. События с одинаковым ключом могут отменять друг друга. На сервер будет отправлено только первое событие и возможно последнее.

Классы событий расширены методом создания PoolEntry. Т.к. классы событий пользователя и чата не связаны между собой, они имеют отдельные расширения. Для создания PoolEntry указывается способ расчета ключа и хеша значимых полей события. 

Вся логика сосредоточена в классе OptimisticEventsPoolService, для подключения события к этому механизму
1) В месте, где клиент обрабатывает событие, обращение к серверу оборачивается в  PoolEntry. 
2) В месте, где происходит определение разбор событий от сервера добавляется проверка на игнорирование события, одна на все типы событий.

SleepySquash commented:
@Alienjob, во-первых, это 100% не сервис, потому что разве это бизнес-логика? В доменной области "мессенджер" нет понятия "сервис событий", это не относится к сервисам. Это деталь реализации стора, только стор об этом и знает. Соответственно, такая структура будет утилитой, но не сервисом. Утилита может создаваться в каждом сторе, в неё будут кормить сторы свои события и действия для последовательной обработки, которую Вы описали.

Во-вторых, актуальность такой структуры в целом звучит адекватно. Нам нужно не просто "не обрабатывать события, которые обработали оптимистично", но и исключить спам таких оптимистичных действий, чтобы на бэкэнд не отправлять миллионы однотипных ивентов - как в звонке, например, при поднятии/опускании руки, если не ошибаюсь. Но этот механизм - это по сути debounce, в `getx` даже есть под желаемое поведение воркер под названием `interval` - он вызывает обработчик не больше, чем раз в указанный период, если что-то было изменено, ну плюс нам ещё нужно первое изменение сразу же кинуть на бэкэнд, без ожиданий. Важно не переизобрести велосипед такой структурой, т.е. её реализация тут играет __ключевую__ роль.

SleepySquash commented:
@Alienjob, т.е. у меня пока претензия к реализации - мне кажется, что она излишне сложная. Но не буду утверждать, что можно проще, буду только просить Вас, пожалуйста, попробовать её упростить максимально сильно.

Alienjob committed:
Revert "Implement dependency injection for event pool"

This reverts commit e4ee0f8e2b180ed7fa9e002cbb030312f11e5181.

Alienjob committed:
Remove global EventPool

Alienjob committed:
Refactor

Alienjob committed:
Refactoring

- remove EventMode
- make toPoolEntry method more kind-centred
- implement == for PoolEntry

Alienjob committed:
Refactor and improve documentation

Alienjob committed:
Improve documentation

Alienjob committed:
Prevent putting chat to Hive after ignored event

Alienjob committed:
Fix imports

Alienjob committed:
Remove debug log

Alienjob committed:
Remove unsupported type of event

Alienjob committed:
Fix import

Alienjob committed:
Make EventPool.add awaitable

Alienjob commented:
**FCM**

```
Implement tracking collapsable client events (#768, #444)

- implement EventPool for collapsable events
- implement convertation from [MyUserEvent] to [PoolEntry] for [muteChatsToggled] type
- implement convertation from [ChatEvent] to [PoolEntry] for [chatFavoriteToggled] type
```

Alienjob commented:
@SleepySquash Постарался максимально упростить реализацию.

Сейчас в ней нет как такового `debounce` - задержки перед вызовом бека. Есть отмена одним событием другого, еще не обработанного. Например, если пока обрабатывается поднятие руки пользователь опустил руку, событие зарегистрируется к обработке. Если он снова поднял ее, пока первый запрос на поднятие не завершился, то новое поднятие отменит обработку опускания и не будет обработано само.

`debounce` может заменить механизм отмены, при этом усложнив его - понадобится дополнительно хранить последнее отправленное значение. Это запретило бы спам на сервер, но данный ПР не пытается решить проблему спама. Он уменьшает спам за счет последовательной обработки однотипных запросов и механизма отмены событий.

Можно было бы дополнительно упростить реализацию если
- Изменить поведение бека - не слать на клиент события, которые уже оптимистично обработаны. Т.е. не слать событие мьюта в сессию, которая его породила. Тогда можно избавиться от механизма игнорирования, останется только контроль последовательного выполнения запросов и отмена дублей.
- Отказаться от механизма отмены - последовательно выполнять все действия пользователя.

Alienjob committed:
Add to EventPool PoolEntry instead of event

Alienjob committed:
Sort PoolEntry constructor

Alienjob committed:
Fix unit test

Alienjob committed:
Merge branch 'main' into 444-track-displace-client-events

Alienjob committed:
Fix changes after merge

Alienjob committed:
Improve documentation

SleepySquash commented on lib/util/event_pool.dart:
```diff
@@ -0,0 +1,184 @@
+  /// Stores hashed valuable properties of event, realted to this [PoolEntry].
+  final int propsHash;
+
+  @override
+  int get hashCode => propsHash;
+
+  @override
+  bool operator ==(Object other) =>
+      other is PoolEntry && propsHash == other.propsHash;
```
Я не уверен, что за сравнения такого рода должен отвечать этот `PoolEntry`. Не должно ли сравнение происходить иначе, т.е. через сравнение внутри объектов или через колбэк `compare` или типа того, как в объектах из стандартной библиотеке (`sort`, всякие `SplayTreeMap`, вроде бы)?

SleepySquash commented on lib/util/event_pool.dart:
```diff
@@ -0,0 +1,184 @@
+/// Extension adding [PoolEntry] construction from a [ChatEvent].
+extension ChatEventToPoolEntryExtension on ChatEvent {
+  /// Constructs a new [PoolEntry] from this [ChatEvent].
+  PoolEntry? toPoolEntry([Future<void> Function()? handler]) {
+    return switch (kind) {
+      ChatEventKind.favorited => PoolEntry(
+          type: EventType.chatFavoriteToggled,
```
Не уверен вовсе, что есть смысл в создании неких `PoolEntry`. Почему мы не можем просто ивенты закидывать напрямую?
